Automattic/newspack-plugin
# feat: ras data library

### All Submissions:






### Changes proposed in this Pull Request:

A reader data library using localStorage. Along with the library implementation, this PR also tweaks the content gate's frontend metering strategy to use it.

### Reader data

Gets and sets reader data using localStorage. Each item is a localStorage entry prefixed with `np_reader_` and a stringified JSON value.

```js
newspackRAS.push( ras => {
  const reader = ras.store.get( 'reader' ); // { "email": "john@doe.com", "authenticated": true }
} );
```

Updating a value emits a `data` event:

```js
newspackRAS.push( ras => {
  ras.on( 'data', ( { detail: { key, value } } ) => {
    console.log( { key, value } );
  } );

  ras.store.set( 'isDonor', true ); // Logs { key: 'isDonor', value: true }
  ras.store.delete( 'isDonor' ); // Logs { key: 'isDonor', value: undefined }
} );
```

#### Data sync

By default, every data update will attempt to sync with the database as user meta. For anonymous sessions, an internal `unsynced` array will store the keys that are pending synchronization until there's an authenticated session that allows the data push.

Items can also skip synchronization on `.set()`:

```js
newspackRAS.push( ras => {
  ras.store.set( 'foo', 'bar', false ); // Will not sync
  ras.store.set( 'foo', 'bar', true ); // Syncs
  ras.store.set( 'foo', 'bar' ); // Syncs
} );
```

For authenticated readers, synced items will rehydrate the store on page load, which allows for systems that rely on the store to have consistent reader data across different sessions.

#### Backend handling of synced data

Using the synchronization strategy, data can be originated and managed on the backend:

The following will update the reader localStorage store on the next page load:
```php
$user_id = 123;
$key     = 'isDonor';
$value   = true;
\Newspack\Reader_Data::update_item( $user_id, $key, wp_json_encode( $value ) );
```

Note that the value should always be stringified JSON, so make sure to always use `wp_json_encode()`. Even for strings so `'test'` becomes the json compliant `'"test"'`

You can also delete and get store items:

```php
$user_id = 123;
$key     = 'isDonor';

$value = \Newspack\Reader_Data::get_data( $user_id, $key );

\Newspack\Reader_Data::delete_item( $user_id, $key );
```

### Reader activity

The store supports "collections", which will be used for managing reader activity with a schema that integrates with [data events](https://github.com/Automattic/newspack-plugin/tree/master/includes/data-events).

The collection can hold up to 1,000 items and each lives for 30 days.

The following code:
```js
newspackRAS.push( [ 'donate', { 'amount': 5 } ] );
```

1. Updates the `np_reader_activity` localStorage entry to include:
     ```json
     { "action": "donate", "data": { "amount": 5 }, "timestamp": 123456789 }
     ```
2. Emits an `activity` event:
     ```js
     newspackRAS.push( ras => {
       ras.on( 'activity', ( { detail: { action, data, timestamp } ) => {
         console.log( 'Reader activity', { action, data, timestamp } );
       }
     } );
     ```

Activities can be fetched with:

```js
newspackRAS.push( ras => {
  const donations = ras.getActivities( 'donate' );
  console.log( 'Donations', donations );

  const allActivities = ras.getActivities();
  console.log( 'All activities', allActivities );
} ); 
```

The next step is to have proper integration with data events: a new activity can dispatch a data event on the backend and a data event can dispatch a reader activity to the frontend.

### How to test the changes in this Pull Request:

1. Confirm the new tests have good coverage
2. In a new session authenticate with a reader account
4. Open devtools and confirm the localStorage contains the reader entry
5. Using the devtools console, update a data item: `newspackReaderActivation.store.set( 'foo', 'bar' );`
6. Confirm the value is added to the store and also quickly added and removed from `np_reader__unsynced` (the queue polling strategy picked it up the synced it to the backend)
7. Visit your database and confirm the user meta is added as `newspack_reader_data_item_foo`
8. Manually remove the item from localStorage, refresh the page, and confirm it's rehydrated back
9. Remove using the API: `newspackReaderActivation.store.delete( 'foo' );`
10. Refresh the page and confirm the value is not added and it has been removed as user meta
11. Start a fresh anonymous session and update several values:
       ```
       newspackReaderActivation.store.set( 'foo', 'baz' );
       newspackReaderActivation.store.set( 'boolean', true );
       newspackReaderActivation.store.set( 'array', [ 1, 2, 3 ] );
       ```
12. Confirm the values have been added to localStorage and the `np_reader__unsynced` item contains `[ 'foo', 'boolean', 'array' ]`
13. Authenticate to that same reader account and navigate to any page (so the API token is available)
14. Confirm the unsynced items are cleared and the user meta is updated with the stringified JSON values
15. `foo` user meta value should update from `bar` to `baz`
16. Test pushing an activity: `newspackRAS.push( [ 'test-activity', { foo: 'bar' } ] );`
17. Confirm the `np_reader_activity` entry is updated to include the activity with a `timestamp` property

### Other information:
